src/nexmark/queries/q12.rs

```diff
@@ -0,0 +1,150 @@
+use super::{process_time, NexmarkStream};
+use crate::{nexmark::model::Event, operator::FilterMap, Circuit, OrdZSet, Stream};
+
+///
+/// Query 12: Processing Time Windows (Not in original suite)
+///
+/// How many bids does a user make within a fixed processing time limit?
+/// Illustrates working in processing time window.
+///
+/// Group bids by the same user into processing time windows of 10 seconds.
+/// Emit the count of bids per window.
+///
+/// ```sql
+/// CREATE TABLE discard_sink (
+///   bidder BIGINT,
+///   bid_count BIGINT,
+///   starttime TIMESTAMP(3),
+///   endtime TIMESTAMP(3)
+/// ) WITH (
+///   'connector' = 'blackhole'
+/// );
+///
+/// INSERT INTO discard_sink
+/// SELECT
+///     B.bidder,
+///     count(*) as bid_count,
+///     TUMBLE_START(B.p_time, INTERVAL '10' SECOND) as starttime,
+///     TUMBLE_END(B.p_time, INTERVAL '10' SECOND) as endtime
+/// FROM (SELECT *, PROCTIME() as p_time FROM bid) B
+/// GROUP BY B.bidder, TUMBLE(B.p_time, INTERVAL '10' SECOND);
+/// ```
+
+type Q12Stream = Stream<Circuit<()>, OrdZSet<(u64, u64, u64, u64), isize>>;
+const TUMBLE_SECONDS: u64 = 10;
+
+fn window_for_process_time(ptime: u64) -> (u64, u64) {
+    let window_lower = ptime - (ptime % (TUMBLE_SECONDS * 1000));
+    (window_lower, window_lower + TUMBLE_SECONDS * 1000)
+}
+
+// This function enables us to test the q12 functionality without using the
+// actual process time, while the actual q12 function below uses the real
+// process time.
+// TODO: I originally planned to pass a FnMut closure for process_time that
+// just emits a new u64 each time it is called, but can't do this as the
+// closure of `flat_map_index` is Fn not FnMut, and would need to capture the
+// process_time closure. So right now, it's quite ugly: to avoid an `FnMut`,
+// I'm instead passing an optional vector of times with the assumption that
+// those times are indexed by the bid.auction.
+// There must be a better way without resorting to interior mutability? Anyway,
+// it works for the tests and is only used in the tests.
```

Comment:
If it's only used in tests where performance doesn't matter, you should be able to get away with `Fn` by relying on internal mutability, by wrapping whatever mutable state the function needs in `Arc<Mutex<>>`.

Reply:
Since each test runs in its own thread, it seems that a `RefCell<>` is enough? Pushed, I'll land if everything passes.